File: java/sdk/src/main/java/network/t0/sdk/network/RequestValidationInterceptor.java
package network.t0.sdk.network;

import build.buf.protovalidate.ValidationResult;
import build.buf.protovalidate.Validator;
import build.buf.protovalidate.ValidatorFactory;
import build.buf.protovalidate.exceptions.ValidationException;
import com.google.protobuf.Message;
import io.grpc.*;
import network.t0.sdk.common.ValidationUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * gRPC client interceptor that validates outgoing requests against
 * buf.validate proto annotations before they are sent.
 *
 * <p>Invalid requests are rejected with {@link Status#INVALID_ARGUMENT}
 * before they leave the client.
 *
 * <p>This class is thread-safe. The {@link Validator} instance is created once
 * and reused across all calls.
 */
public final class RequestValidationInterceptor implements ClientInterceptor {

    private static final Logger log = LoggerFactory.getLogger(RequestValidationInterceptor.class);

    private final Validator validator;

    public RequestValidationInterceptor() {
        this.validator = ValidatorFactory.newBuilder().build();
    }

    @Override
    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
            MethodDescriptor<ReqT, RespT> method,
            CallOptions callOptions,
            Channel next) {
        return new ForwardingClientCall.SimpleForwardingClientCall<>(next.newCall(method, callOptions)) {
            @Override
            public void sendMessage(ReqT message) {
                if (message instanceof Message protoMessage) {
                    try {
                        ValidationResult result = validator.validate(protoMessage);
                        if (!result.isSuccess()) {
                            String details = ValidationUtils.formatViolations(result);
                            log.error("Request validation failed for {}: {}", protoMessage.getDescriptorForType().getFullName(), details);
                            throw Status.INVALID_ARGUMENT.withDescription("request validation failed: " + details).asRuntimeException();
                        }
                    } catch (ValidationException e) {
                        log.error("Request validation error for {}: {}", protoMessage.getDescriptorForType().getFullName(), e.getMessage());
                        throw Status.INVALID_ARGUMENT.withDescription("request validation error: " + e.getMessage()).asRuntimeException();
                    }
                }
                super.sendMessage(message);
            }
        };
    }
}

File: csharp/sdk/T0.ProviderSdk/Network/RequestValidationInterceptor.cs
using Grpc.Core;
using Grpc.Core.Interceptors;
using Google.Protobuf;
using ProtoValidate;
using T0.ProviderSdk.Common;

namespace T0.ProviderSdk.Network;

/// <summary>
/// gRPC client interceptor that validates outgoing requests against
/// buf.validate proto annotations before they are sent.
///
/// Invalid requests are rejected with StatusCode.InvalidArgument before leaving the client.
/// </summary>
public sealed class RequestValidationInterceptor : Interceptor
{
    private static readonly Validator _validator = new();

    public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(
        TRequest request,
        ClientInterceptorContext<TRequest, TResponse> context,
        AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
    {
        if (request is IMessage message)
        {
            var result = _validator.Validate(message, failFast: false);
            if (!result.IsSuccess)
            {
                throw new RpcException(new Status(StatusCode.InvalidArgument,
                    $"request validation failed: {ValidationUtils.FormatViolations(result)}"));
            }
        }

        return continuation(request, context);
    }

}
